app/modules/cart/usecases/create_cart.py

```diff
@@ -1,20 +1,104 @@
+    DuplicateEntryException,
+    EntityNotFoundException,
+)
 from app.modules.cart.repository_interface import CartItemRepositoryInterface
+from app.modules.product.repository_interface import ProductRepositoryInterface
 
-from ..schemas import CartBulkCreate, CartRead
+from ..schemas import CartBulkCreate, CartItemRead, CartRead
 
 
 class CreateCartFromItems:
-    def __init__(self, cart_item_repository: CartItemRepositoryInterface) -> None:
+    def __init__(
+        self, cart_item_repository: CartItemRepositoryInterface, product_repository: ProductRepositoryInterface
+    ) -> None:
         self.cart_item_repository = cart_item_repository
+        self.product_repository = product_repository
 
     async def execute(self, cart_bulk: CartBulkCreate) -> CartRead | None:
         """
         Create a cart from a list of items.
-        This method will insert multiple cart items to the database with repository bulk_insert method.
-        If any item already exists for the user, it will raise a DuplicateEntryException.
-        If the insertion fails, it will raise a DatabaseOperationException.
-        The return value is a CartRead object containing the list of CartItemRead objects.
-        If the cart is empty, it will return None.
+
+        Steps:
+        1. Validate that cart_bulk is not empty.
+        2. Check for duplicate items (user_id + product_id) in a single query.
+        3. Map input DTOs to repository models.
+        4. Bulk insert into the database with exception handling.
+        5. Map inserted items to CartItemRead DTOs.
+        6. Return a CartRead containing all items.
+
+        Raises:
+            DuplicateEntryException: if any item already exists for the user.
+            DatabaseOperationException: if bulk insert fails.
+
+        Returns:
+            CartRead | None: Returns None if cart is empty.
         """
 
-        pass
+        if not cart_bulk.items:
+            return None
+
+        for item in cart_bulk.items:
+            product = await self.product_repository.get_by_id(item.product_id)
+            if not product:
+                raise EntityNotFoundException(
+                    data={"product_id": item.product_id}, message=f"Product with id {item.product_id} not found"
+                )
+
+        # Prepare list of (user_id, product_id) tuples
+        user_product_pairs: List[Tuple[int, int]] = [(item.user_id, item.product_id) for item in cart_bulk.items]
+
+        # Query existing items in one go
+        stmt = select(
+            self.cart_item_repository.model_class.user_id, self.cart_item_repository.model_class.product_id
+        ).where(
+            tuple_(self.cart_item_repository.model_class.user_id, self.cart_item_repository.model_class.product_id).in_(
+                user_product_pairs
+            )
+        )
+
+        result = await self.cart_item_repository.session.execute(stmt)
+        existing_pairs = set(result.fetchall())
+
+        if existing_pairs:
```

Comment:
Error on First Duplicate Only: Only the first duplicate pair is reported, potentially missing others.